File: vader/src/main/java/org/revcloud/vader/config/FieldConfig.java
/*******************************************************************************
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 ******************************************************************************/

/*******************************************************************************
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 ******************************************************************************/

package org.revcloud.vader.config;

import de.cronn.reflection.util.TypedPropertyGetter;
import io.vavr.Function2;
import io.vavr.Tuple2;
import java.util.Collection;
import java.util.Map;
import java.util.function.Predicate;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.NonNull;
import lombok.Singular;
import lombok.Value;
import lombok.experimental.FieldDefaults;
import org.jetbrains.annotations.Nullable;

@Value
@FieldDefaults(level = AccessLevel.PACKAGE)
@Builder(buildMethodName = "prepare", builderMethodName = "toValidate", toBuilder = true)
public class FieldConfig<FieldT, ValidatableT, FailureT> {
  @Nullable Predicate<FieldT> withFieldValidator;

  @Singular("shouldHaveValidFormatOrFailWith")
  Map<TypedPropertyGetter<ValidatableT, FieldT>, FailureT> shouldHaveValidFormatForAllOrFailWith;

  @Nullable
  Tuple2<
          @NonNull Collection<@NonNull TypedPropertyGetter<ValidatableT, @Nullable FieldT>>,
          @NonNull Function2<String, @Nullable FieldT, @Nullable FailureT>>
      shouldHaveValidFormatForAllOrFailWithFn;

  @Singular("shouldHaveValidFormatOrFailWithFn")
  Map<TypedPropertyGetter<ValidatableT, FieldT>, Function2<String, FieldT, FailureT>>
      shouldHaveValidFormatOrFailWithFn;

  @Singular("absentOrHaveValidFormatOrFailWith")
  Map<TypedPropertyGetter<ValidatableT, FieldT>, FailureT> absentOrHaveValidFormatForAllOrFailWith;

  @Nullable
  Tuple2<
          @NonNull Collection<@NonNull TypedPropertyGetter<ValidatableT, @Nullable FieldT>>,
          @NonNull Function2<String, @Nullable FieldT, @Nullable FailureT>>
      absentOrHaveValidFormatForAllOrFailWithFn;

  @Singular("absentOrHaveValidFormatOrFailWithFn")
  Map<TypedPropertyGetter<ValidatableT, FieldT>, Function2<String, FieldT, FailureT>>
      absentOrHaveValidFormatOrFailWithFn;
}

File: vador/src/main/kotlin/com/salesforce/vador/config/FieldConfig.kt
/**
 * ****************************************************************************
 * Copyright (c) 2022, salesforce.com, inc. All rights reserved. SPDX-License-Identifier:
 * BSD-3-Clause For full license text, see the LICENSE file in the repo root or
 * https://opensource.org/licenses/BSD-3-Clause
 * ****************************************************************************
 */
package com.salesforce.vador.config

import com.salesforce.vador.immutables.ConfigStyle
import de.cronn.reflection.util.TypedPropertyGetter
import io.vavr.Function2
import io.vavr.Tuple2
import java.util.function.Predicate
import org.immutables.value.Value
import org.jetbrains.annotations.Nullable

@ConfigStyle
@Value.Immutable(copy = false)
internal abstract class AbstractFieldConfig<FieldT, ValidatableT, FailureT> {
  @get:Nullable abstract val withFieldValidator: Predicate<FieldT>?

  @get:Value.Auxiliary
  protected abstract val shouldHaveValidFormatForAllOrFailWiths:
    Map<TypedPropertyGetter<ValidatableT, FieldT>, FailureT>

  @get:Value.Derived
  open val shouldHaveValidFormatForAllOrFailWith:
    Map<TypedPropertyGetter<ValidatableT, FieldT>, FailureT>
    get() = shouldHaveValidFormatForAllOrFailWiths

  @get:Nullable
  abstract val shouldHaveValidFormatForAllOrFailWithFn:
    Tuple2<
      Collection<TypedPropertyGetter<ValidatableT, FieldT?>>,
      Function2<String, FieldT?, FailureT?>,
    >?

  @get:Value.Auxiliary
  protected abstract val shouldHaveValidFormatOrFailWithFns:
    Map<TypedPropertyGetter<ValidatableT, FieldT>, Function2<String, FieldT, FailureT>>

  @get:Value.Derived
  open val shouldHaveValidFormatOrFailWithFn:
    Map<TypedPropertyGetter<ValidatableT, FieldT>, Function2<String, FieldT, FailureT>>
    get() = shouldHaveValidFormatOrFailWithFns

  @get:Value.Auxiliary
  protected abstract val absentOrHaveValidFormatForAllOrFailWiths:
    Map<TypedPropertyGetter<ValidatableT, FieldT>, FailureT>

  @get:Value.Derived
  open val absentOrHaveValidFormatForAllOrFailWith:
    Map<TypedPropertyGetter<ValidatableT, FieldT>, FailureT>
    get() = absentOrHaveValidFormatForAllOrFailWiths

  @get:Nullable
  abstract val absentOrHaveValidFormatForAllOrFailWithFn:
    Tuple2<
      Collection<TypedPropertyGetter<ValidatableT, FieldT?>>,
      Function2<String, FieldT?, FailureT?>,
    >?

  @get:Value.Auxiliary
  protected abstract val absentOrHaveValidFormatOrFailWithFns:
    Map<TypedPropertyGetter<ValidatableT, FieldT>, Function2<String, FieldT, FailureT>>

  @get:Value.Derived
  open val absentOrHaveValidFormatOrFailWithFn:
    Map<TypedPropertyGetter<ValidatableT, FieldT>, Function2<String, FieldT, FailureT>>
    get() = absentOrHaveValidFormatOrFailWithFns

  abstract class Builder<FieldT, ValidatableT, FailureT> :
    FieldConfigBuilderDsl<
      FieldT,
      ValidatableT,
      FailureT,
      FieldConfig.Builder<FieldT, ValidatableT, FailureT>,
    >
}
